feat(custom-date-picker): updated custom date, time, date range pickers w flatpickr

wbarbee opened the pull request:
## Summary

Refactors **datepicker**, **timepicker**, **date range picker** components, utilizing [`flatpickr`](https://flatpickr.js.org/) third-party library. 

## ADO Story or GitHub Issue Link

[UI: Datepicker User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1910721)
[UI: Date/time range picker User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1910724)
[UI: Timepicker User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1910722)
[UI: Flatpickr theme User Story](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1937059)

## Figma Link
[https://www.figma.com/design/dhPuQQrqxHqMvtmnMMuTry/Santorini---Global-Filters?node-id=518-47488&node-type=frame&m=dev](https://www.figma.com/design/dhPuQQrqxHqMvtmnMMuTry/Santorini---Global-Filters?node-id=518-47488&node-type=frame&m=dev)

## Notes

**primary considerations**:
- validation -- flatpickr out of the box/plugins vs. dropping all validations except for `invalidText`
- custom shidoka theming

## To Do

- potential 2.0 breaking changes **chore** -- restructure Storybook side nav items for these three components, possibly nesting them under a common folder (e.g., `dateTimePicker`, `Date Time Picker`)

## Checklist

- [x] Used Conventional Commit messages as outlined in the contributing guide.
  - [x] Noted breaking changes (if any).
- [x] Documented/updated all props, events, slots, parts, etc with JSDoc.
  - [x] Ran the `analyze` command to update Storybook docs.
- [x] Added/updated Stories with controls to cover all variants.
- [x] Ran `test` locally to address any failures.
- [x] Added/updated the Figma link for the Story's Design tab.
- [x] Added any new component exports to the src/index.ts file

## Screenshots

<img width="1120" alt="Screenshot 2024-10-16 at 9 45 04 AM" src="https://github.com/user-attachments/assets/36ae43ee-c0de-4f8b-a883-5d191e33d81d">

<img width="1237" alt="Screenshot 2024-10-16 at 4 51 30 PM" src="https://github.com/user-attachments/assets/2d5e28cf-e0f5-4658-ba17-f248c4474224">

<img width="1149" alt="Screenshot 2024-10-16 at 9 45 51 AM" src="https://github.com/user-attachments/assets/fd201ba3-f019-4b97-9286-29bbf9f8f468">

<img width="1166" alt="Screenshot 2024-10-16 at 9 46 54 AM" src="https://github.com/user-attachments/assets/8ece108a-ed72-4cb1-8e75-76b945677d3c">

wbarbee commented on .gitignore:
```diff
@@ -11,3 +11,5 @@ coverage
 # logs
 .eslintcache
```
this is a macOS thing. i still don't fully understand what triggers it being included in a commit, but adding to the .gitignore as it does not include any project-relevant or project-critical info

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,217 @@
+      ?required=${args.required}
+      aria-invalid=${args._isInvalid ? 'true' : 'false'}
+    />
+    <span slot="icon" style="line-height: 1;">${unsafeSVG(overflowIcon)}</span>
```
Should this be a calendar icon, why overflow?

wbarbee replied:
oh, yes this should be a calendar icon

wbarbee replied:
fixed this in migrating inputs from slot to internal

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,217 @@
+        : null}
+      ${args.unnamed}
+    </label>
+    <kd-button
```
A button anchor concept is interesting, but I don't understand how it can function correctly with a form if there is no `<input>`, and how the user is supposed to know the selected date without opening the picker.

wbarbee replied:
button options removed in datepicker, timepicker, daterangepicker, starting with an internal, single input element that fills 100% of parent/containing element

i can implement sizing (sm, md, lg) if we want to do that in this case.

brian-patrick-3 commented on src/components/reusable/daterangepicker/daterangepicker.stories.js:
```diff
@@ -1,122 +1,420 @@
+  `;
+};
+
+const MultiInputTemplate = (args) => {
```
I think we just need to go with one variant, probably single input.

wbarbee replied:
✅

brian-patrick-3 reviewed (commented):
There's a lot going on in here, so I just left a few high level comments on things that jumped out at me for now.

wbarbee commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,217 @@
+    .maxDate="${args.maxDate}"
+    @on-change=${(e) => action(e.type)(e)}
+  >
+    <label
+      slot="label"
+      class="label-text"
+      for=${anchorId}
+      ?disabled=${args.datePickerDisabled}
+    >
+      ${args.required
+        ? html`<abbr
+            class="required"
+            title=${args._textStrings?.requiredText || 'Required'}
+            aria-label=${args._textStrings?.requiredText || 'Required'}
+            >*</abbr
+          >`
+        : null}
+      ${args.unnamed}
+    </label>
+    <input
+      slot="anchor"
+      type="text"
+      id=${anchorId}
+      name=${args.nameAttr}
+      placeholder=${placeholder}
+      ?disabled=${args.datePickerDisabled}
+      ?required=${args.required}
+      aria-invalid=${args._isInvalid ? 'true' : 'false'}
+    />
+    <span slot="icon" style="line-height: 1;">${unsafeSVG(overflowIcon)}</span>
```
for this comment and your 3rd comment about the overflow icon, i was working off of the implementation that they have in the figma file where it would be part of a global filter or something.

i suppose we could just have two options -- input or button, but i originally had it slotted instead of internal for flexibility of the anchor. let me know what you think re: having 2-3 hardcoded internal options, just input, etc.

brian-patrick-3 replied:
I see that in the old design now. I think what the design is getting at there, is they just want some sort of popup menu to house the datepicker in that scenario. I think that design is not well thought out. The calendar selection should probably still be tied to an actual input, and that's how all of the flatpickr examples work. Even the [inline calendar example](https://flatpickr.js.org/examples/#inline-calendar) still shows an input, though maybe there could be a way to hide it.

I think the most important thing is that everything still works with forms. Slotted elements are also exposed to external styling, since they live in the light dom, something to consider.

wbarbee replied:
understood. yeah it's been a little difficult to get the design nailed down on this one. still working to get better about communicating the need to have things more clearly defined. 

in the meantime, i've tried to pack in features that i can rip out when the time comes, which is overkill. but easy to scale it back to more internal, simpler variations.

i'll take another stab with all of this in mind and see if we can get it closer

wbarbee replied:
ok this should be a bit more on track. i've set `required` to true in storybook for each component by default, as that's what initializes the validation and it will be easier to test. 

two behaviors that i've noticed with flatpickr:
1) for timepicker, if you click to open the timepicker and do not 'touch it' or modify the time at all, then close it by clicking anywhere outside of the overlay, there will be a moment where it does my standard validation and recognizes that the field is required, that no time has been chosen. however, i think flatpicker then defaults to 12:00

2) on the daterangepicker, when you select only one date, my validations are working as expected. however, when you open the datepicker overlay again, it has reset to no dates selected. this is more of a standard date picker behavior, but just wanted to note that as well.

wbarbee commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,217 @@
+    >
+      ${args.required
+        ? html`<abbr
+            class="required"
```
this has been fixed by moving majority of slots internal, minus the label unnamed slot

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,116 @@
+    .maxDate="${args.maxDate}"
+    @on-change=${(e) => action(e.type)(e)}
+  >
+    ${args.unnamed}
```
Part of the 2.0 changes has some [chores for the form components](https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1931303/), can you please include those changes to these components as well? I removed them from the tasks for that story so you don't get conflicts.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,116 @@
+    ?required="${args.required}"
+    .value="${args.value}"
+    .warnText="${args.warnText}"
+    .invalidText="${args.invalidText}"
```
I'm not seeing the custom error message appear under the component when I update this control.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -79,135 +146,295 @@ export class DatePicker extends FormMixin(LitElement) {
-          ? html`<div id="warning" class="warn">${this.warnText}</div>`
-          : null}
+
+        ${this.caption
```
When I click on the caption or the error message, it also opens the picker.

wbarbee replied:
this one is a bit elusive. i don't know if this is a weird side-effect of flatpickr as well. i've tried adjusting the html and pointer-events but not currently working as expected. still working on it

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -79,135 +146,295 @@ export class DatePicker extends FormMixin(LitElement) {
-          : null}
+
+        ${this.caption
+          ? html`<div id=${descriptionId} class="caption options-text">
```
Caption styles don't seem to match the other form components, it should be common through the form-input scss

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,116 @@
+    .dateFormat="${args.dateFormat}"
+    ?required="${args.required}"
+    .value="${args.value}"
+    .warnText="${args.warnText}"
```
When I add warning text is appears twice.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.stories.js:
```diff
@@ -1,114 +1,116 @@
+    .nameAttr="${args.nameAttr}"
+    .locale="${args.locale}"
+    .dateFormat="${args.dateFormat}"
+    ?required="${args.required}"
```
If you update the `patterns/forms` page with these components set as required, it will be easier to test the validation with form submit.

wbarbee replied:
✅

brian-patrick-3 commented on src/common/scss/form-input.scss:
```diff
@@ -79,6 +80,15 @@ textarea {
   margin-top: 8px;
 
+  &.error-text {
+    font-size: 16px;
```
Shouldn't need to specify this font size, it's the default body font size.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.scss:
```diff
@@ -1,28 +1,135 @@
 :host {
-  display: inline-block;
+    display: inline-block;
+    width: 100%;
```
We don't set a default width on any of the other form inputs, usually just let the application control it.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/timepicker/timepicker.stories.js:
```diff
@@ -1,76 +1,91 @@
+      .nameAttr="${args.nameAttr}"
+      .locale="${args.locale}"
+      ?required="${args.required}"
+      .value="${args.value}"
```
When I set 12:00 AM on the timepicker, it appears as `00:00`

wbarbee replied:
this is a flatpickr thing, i think. you can test here:

https://flatpickr.js.org/examples/#time-picker

brian-patrick-3 replied:
I see, interesting default. I wonder if we can [change the default format](https://flatpickr.js.org/formatting/#time-formatting-tokens) so it shows as "12:00 AM"? Actually, I don't see a prop on the timepicker component to modify the format like datepicker has.

wbarbee replied:
from what i could tell from the docs, the dateFormat doesn't behave quite the same for Timepicker as it does for Datepicker. i tried another time format like `H:i:s` and it didn't work.

that's why i'm hardcoding that value, but i can try again to make it work and get the 12H format to show in the input

brian-patrick-3 replied:
I wonder if something like `h:i K` is what we want there for the 12 hour format default.

wbarbee replied:
that works, nice 👍

wbarbee replied:
✅

brian-patrick-3 reviewed (commented):
A couple more things:

1. The validations don't appear to be working with the form, it's treating the form as valid even without entering values for any of the date/time inputs. Take a look at the console on the forms page when clicking submit.
   1. I think you are just missing some of the pieces in the `_validate` function that the other form components have, ex setting validity on the `_internals`.
2. I don't see any migration notes yet. I'm curious what exactly the breaking changes are. Ex: is there a new format for `value`? is locale is no longer automatic? etc

wbarbee commented:
> A couple more things:
> 
> 1. The validations don't appear to be working with the form, it's treating the form as valid even without entering values for any of the date/time inputs. Take a look at the console on the forms page when clicking submit.
>    
>    1. I think you are just missing some of the pieces in the `_validate` function that the other form components have, ex setting validity on the `_internals`.
> 2. I don't see any migration notes yet. I'm curious what exactly the breaking changes are. Ex: is there a new format for `value`? is locale is no longer automatic? removed/changed props? etc

i amended a comment above. still having some issues with the validations, but it's nearly there. aiming to get it sorted tomorrow (mon 21st)

brian-patrick-3 commented on src/stories/migration.mdx:
```diff
@@ -26,6 +26,22 @@ Breaking Changes
     - `label` slot has been updated to a `label` string prop
+- Updated date, time, and date range picker components
+  - Utilizes Flatpickr third-party date picker library to standardize cross-browser date, time, and date range pickers, and to support theming and locale
+  - The locale can now be set via the `locale` attribute
```
Should this be "can" or "must"?

wbarbee replied:
✅ -- totally reorganized this migration guide to be more of a checklist for devs

brian-patrick-3 commented on src/stories/migration.mdx:
```diff
@@ -26,6 +26,22 @@ Breaking Changes
+- Updated date, time, and date range picker components
+  - Utilizes Flatpickr third-party date picker library to standardize cross-browser date, time, and date range pickers, and to support theming and locale
+  - The locale can now be set via the `locale` attribute
+  - The date output format defaults to ISO 8601 but is configurable via the `dateFormat` flatpickr attribute
```
This is visual output right? Should we mention anything about the code input/output formats?

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -1,73 +1,161 @@
-  /** Datepicker size. "sm", "md", or "lg". */
+  /** Sets datepicker attribute name (ex: `contact-form-date-picker`). */
   @property({ type: String })
-  size = 'md';
```
What happened to `size`?

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -1,73 +1,161 @@
+  /** Sets datepicker attribute name (ex: `contact-form-date-picker`). */
   @property({ type: String })
-  size = 'md';
+  nameAttr = '';
```
Should this be `name` to remain consistent with other form components?

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -79,135 +167,327 @@ export class DatePicker extends FormMixin(LitElement) {
+        id=${warningId}
+        class="warn warn-text"
+        role="alert"
+        aria-label=${this.warningAriaLabel || 'Warning message'}
```
We shouldn't need an aria-label if there is visible text. aria-label is really just for interactive elements with no visible text.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -79,135 +167,327 @@ export class DatePicker extends FormMixin(LitElement) {
+        id=${errorId}
+        class="error error-text"
+        role="alert"
+        aria-label=${this.errorAriaLabel || 'Error message'}
```
This aria-label is probably only needed on the icon, since the div also has visible text.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -79,135 +167,327 @@ export class DatePicker extends FormMixin(LitElement) {
+      : this.value;
+
+    emitValue(this, 'on-change', {
+      dates: formattedDates,
+      dateString: (this._inputEl as HTMLInputElement)?.value || dateStr,
```
We should mention the event output structure has changed.

wbarbee replied:
✅

brian-patrick-3 commented on src/components/reusable/datePicker/datepicker.ts:
```diff
@@ -79,135 +167,327 @@ export class DatePicker extends FormMixin(LitElement) {
+      : this.value;
+
+    emitValue(this, 'on-change', {
+      dates: formattedDates,
```
I notice this is a string for a single date, but an array of Date objects for a multi-select. Should the type be consistent (string vs date obj)?

wbarbee replied:
corrected -- handling this as a Date object or an array of Date objects, depending on single vs. multiple `mode`.

brian-patrick-3 reviewed (commented):
I notice some console errors on load from the timepicker and daterangepicker

brian-patrick-3 commented:
LGTM! Once QA approves, you can squash merge and include the `BREAKING CHANGE:` footer notes in the extended description textarea.

wbarbee commented:
fixed the following:

1) minDate / maxDate: when these values are set to the same year (ex: 2024), Flatpickr hides the year from the date header in each calendar
**_FIX_:**  we are now detecting this and removing year element entirely when minDate and maxDate are explicitly set and are of the same year -- otherwise, the year will show as normal
*** 2) **AM/PM bug**: Flatpickr applies locale translation to AM/PM ONLY when date picker and timepicker are standalone components.
**_FIX_**: all non-english locales will default to 24H mode (no AM/PM) unless the `twentyFourHourFormat` boolean is explicitly set to true or false for those locales (otherwise undefined). for all english locales, it will default to 12H format
3) changed example hindi label to english to prevent confusion for devs so that they don't assume this value is being translated via the `locale` value. label is a string property and is to be set by developers, not shidoka's responsibility
4) added reset button to all 3 components (date picker, time picker, date range picker)
5) added multi-select date picker example on the `patterns/forms` page
6) corrected labels that read `Date / Time` to read `Date + Time`

attn: @brian-patrick-3  -- please see item 2 above and let me know your thoughts

great work and thanks for your help @srpriyankashetty !!

srpriyankashetty reviewed (approved):
1. Verified the new Date picker, Date range picker and Time picker changes and looks good
2. Verified that Required error message is displayed for all the pickers
3. Verified the min max range for all the pickers
4. Verified the range selections
5. Verified that Storybook >> Forms has all the above listed picker
6. Verified the date pickers and time pickers in Chrome, Edge and Firefox browser and looks fine
7. Verified the time for 12hrs and 24hrs format
8. Verified setting the locale and language changes in the calendar accordingly

Verified bugs list
Reset/clear button when a value is selected in all the pickers - Fixed
Multi select date picker is missing in the form page - Fixed
The label to 'Date / Time' must be changed to 'Date + Time' in form - Fixed
AM and PM is not changing based on the Locale set - set the time to 24hrs format
When there is Min and Max date range set, the year is not showing in the calendar. Also shows some arrows beside the month - Fixed

brian-patrick-3 commented:
:tada: This PR is included in version 2.0.0-next.10 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/@kyndryl-design-system/shidoka-applications/v/2.0.0-next.10)
- [GitHub release](https://github.com/kyndryl-design-system/shidoka-applications/releases/tag/v2.0.0-next.10)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

brian-patrick-3 commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@kyndryl-design-system/shidoka-applications/v/2.0.0)
- [GitHub release](https://github.com/kyndryl-design-system/shidoka-applications/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
